### DISCUSSION TITLE
Cannot resolve symbol 'SpiChipSelect'

### DISCUSSION THREAD

**Discussion by HeroDesign** [Pi4J/pi4j#375] [General]
Hello!
I'm upgrading a project from v1 to v2. Using pi4j-core 2.6.1

I have the following code:
.....
import com.pi4j.Pi4J;
import com.pi4j.context.Context;
import com.pi4j.io.spi.Spi;
import com.pi4j.io.spi.SpiConfig;
import com.pi4j.io.spi.SpiProvider;
import com.pi4j.io.spi.SpiChipSelect;

        pi4j = Pi4J.newAutoContext();
        SpiConfig spiConfig = Spi.newConfigBuilder(pi4j)
                .id("SPI0")
                .bus(0)
                .chipSelect(SpiChipSelect.CS0)
                .baud(30000000)
                .mode(3)
                .build();

but tin import of import com.pi4j.io.spi.SpiChipSelect; wont resolve. 

Any advice?
Thanks

**Comment by taartspi** [comment]
I altered my devices project to use SNAPSHOT 2.6.1 and the import com.pi4j.io.spi.SpiChipSelect worked and the chip select usage in my program compiled.   
I use intellij and I can easily verify the external libraries and examine them if needed.   Intellij requires a step to force the POM requirements be reprocessed in an existing project.   
Likely you already did the following:    I would re-verify I correctly updated the POM file,  and the correct jars were loaded.  Had I used a manual process to download the jars verify that step. Then do a clean and build   mvn clean install   

If you build using something other than maven, state what you use and maybe some user has a suggestion.

**Reply by HeroDesign** [reply]
Fantastic thanks for the reply.
I'm using IntelliJ and Maven as well.
How do I SNAPTHOT version? I can see the SpiChipSelect in the jar file that maven is pulling? I've done many mvn clean installs

thanks so much

**Comment by taartspi** [comment]
SNAPSHOT version,  my words were confusing,  I modified the POM to use <version>2.6.1-SNAPSHOT</version>   
 Assuming you build using intellij menus:
The next thing I would try is under the intellij File menu     Invalidate caches  -----   select Clear file system cache and local history.    Push Invalidate Cache and Restart     

If the imports are still failing, for information  in the file enter the line: 
import com.pi4j.               and after the pi4j. leave the cursor there and see what next levels of the packages are listed. 


There something simple missing but what it is hasn't hit me, thus the Invalidate steps.

**Reply by HeroDesign** [reply]
Thanks so much again.
I've tried invalidating the cache several times while trying to fix this

I've also added the jar from the maven directory as a dependency in in the module under Project Structure
 
I can import everything else the code needs from com.pi4j.io.spi except SpiChipSelect
I updated pom.xml to be:
```
        <dependency>
            <groupId>com.pi4j</groupId>
            <artifactId>pi4j-core</artifactId>
            <version>2.6.1-SNAPSHOT</version>
        </dependency>
```
but I get the maven error: Cannot resolve com.pi4j:pi4j-plugin-raspberrypi:2.6.1-SNAPSHOT

I also get the following errors, not sure if they are related:
 
Exception in thread "main" com.pi4j.provider.exception.ProviderNotFoundException: Pi4J provider [digital-input] could not be found.  Please include this 'provider' JAR in the classpath.

com.pi4j.provider.exception.ProviderNotFoundException: Pi4J provider [i2c] could not be found.  Please include this 'provider' JAR in the classpath.

com.pi4j.provider.exception.ProviderNotFoundException: Pi4J provider [i2c] could not be found.  Please include this 'provider' JAR in the classpath.

**Reply by HeroDesign** [reply]
Was able to get the SNAPTSHOT to resolve in maven by adding:
`<repositories>
    <repository>
        <id>sonatype-snapshots</id>
        <url>https://oss.sonatype.org/content/repositories/snapshots/</url>
        <snapshots>
            <enabled>true</enabled>
        </snapshots>
    </repository>
</repositories>`

But SpiChipSelect still will not resolve even after mvn clean install and invalidating caches and restart. Also ensured the module has the SNAPSHOT version as a dependency as both the snapshot JAR and the maven JAR, neither resolve. And I can see the SpiChipSelect class in the maven JAR

**Reply by HeroDesign** [reply]
Also now that I have the SNAPSHOT in maven com.pi4j:pi4j-plugin-raspberypi:2.6.1, com.pi4j:pi4j-gpio-extension:1.2 and com.pi4j:pi4j-plugin-;inuxfs:2.6.1 wont resolve in maven because the conflict with the SNAPSHOT

**Comment by taartspi** [comment]
Earlier you mentioned provider [digital-input] and [i2c] not found    I believe you need to fix the provider names,  examples  it is not i2c but rather  pigpio-i2c        Others pigpio-spi pigpio-digital-output pigpio-digital-input   

Starting in 2.5.0 the new providers like  Gpiod linuxfs are loaded.  If you want to stay with the pigpio provider change your code to the following when creating the context   
```java  
 var piGpio = PiGpio.newNativeInstance();
        var pi4j = Pi4J.newContextBuilder().
            noAutoDetect().
            add(PiGpioSpiProvider.newInstance(piGpio)).
            add(PiGpioI2CProvider.newInstance(piGpio)).
            add(PiGpioDigitalOutputProvider.newInstance(piGpio)).
            add(PiGpioDigitalInputProvider.newInstance(piGpio)).
            build();


        System.out.println("-------------------------------------------------");
        System.out.println("PI4J PROVIDERS");
        System.out.println("-------------------------------------------------");
        pi4j.providers().describe().print(System.out);
        System.out.println("-------------------------------------------------");  
```  
Not sure what you are describing by a snapshot conflict.  I will paste the dependencies from my pom.  I have more than you will use.   
All pi4j jars must be the same snapshot.  In this conversation 2.6.1 
```  
  <dependencies>
        <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-api</artifactId>
            <version>${slf4j.version}</version>
        </dependency>
        <!-- change to use the default slf4 simple -->
        <dependency>
            <groupId>com.pi4j</groupId>
            <artifactId>pi4j-plugin-gpiod</artifactId>
            <version>${pi4j.version}</version>
        </dependency>
         <dependency>
                <groupId>net.java.dev.jna</groupId>
                <artifactId>jna</artifactId>
                <version>5.12.1</version>
            </dependency>
         <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-simple</artifactId>
            <version>${slf4j.version}</version>
        </dependency>

        <!-- include Pi4J Core -->
        <dependency>
            <groupId>com.pi4j</groupId>
            <artifactId>pi4j-core</artifactId>
            <version>${pi4j.version}</version>
        </dependency>
        <dependency>
            <groupId>com.pi4j</groupId>
            <artifactId>pi4j-plugin-pigpio</artifactId>
            <version>${pi4j.version}</version>
        </dependency>
        <!--- log4j version 2   -->

        <!-- include Pi4J Plugins (Platforms and I/O Providers) -->
        <dependency>
            <groupId>com.pi4j</groupId>
            <artifactId>pi4j-plugin-raspberrypi</artifactId>
            <version>${pi4j.version}</version>
        </dependency>
        <dependency>
            <groupId>org.jetbrains</groupId>
            <artifactId>annotations</artifactId>
            <version>RELEASE</version>
            <scope>compile</scope>
        </dependency>
        <dependency>
            <groupId>com.pi4j</groupId>
            <artifactId>pi4j-plugin-linuxfs</artifactId>
            <version>${pi4j.version}</version>
        </dependency>
    </dependencies>
```
And still interested what is listed when you:  
If the imports are still failing, for information in the file enter the line:
import com.pi4j. and after the pi4j. leave the cursor there and see what next levels of the packages are listed.

**Comment by HeroDesign** [comment]
Thanks so much. for your continued support!

I've been able to get everything loaded with maven thanks to you example pom.

in my SPI code I still can not resolve com.pi4j.io.spi.SpiChipSelect

When I enter the line import com.pi4j. and wait I see the following:
io.
plugin.
context.
Pi4j
common.
boardinfo.
config.
event.
exception.
extension.
internal.
library.
platform.
provider.
registry.
runtime.
util.

I've been using the example here: https://www.pi4j.com/documentation/io-examples/i2c/ for my i2c code. And have the following:
```
    public final static int rtcAddy = 0x51;
    private static I2C rtc = null;
    public RTC() {
        try {
            // Create Pi4J context (this automatically loads the necessary providers)
            Context pi4j = Pi4J.newAutoContext();
            I2CProvider i2CProvider = pi4j.provider("linuxfs-i2c");

            // Create I2C configuration
            I2CConfig config = I2C.newConfigBuilder(pi4j)
                    .bus(1)        // Use I2C bus 1
                    .device(rtcAddy)  // Set the I2C device address
                    .id("RTC")     // Set an ID for the device
                    .name("RTC Device")  // Set a name for the device
                    .build();
            rtc = i2CProvider.create(config);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
```
but get the following error:
SLF4J(W): No SLF4J providers were found.
SLF4J(W): Defaulting to no-operation (NOP) logger implementation
SLF4J(W): See https://www.slf4j.org/codes.html#noProviders for further details.
Exception in thread "main" java.lang.UnsatisfiedLinkError: Pi4J was unable to extract and load the native library [/lib/armhf/pi4j-gpiod/libgpiod.so] from the embedded resources inside this JAR [/home/hero/.m2/repository/com/pi4j/pi4j-library-gpiod/2.6.1-SNAPSHOT/pi4j-library-gpiod-2.6.1-20240729.195405-3.jar]. to a temporary location on this system.  You can alternatively define the 'pi4j.library.path' system property to override this behavior and specify the library path.
 UNDERLYING EXCEPTION: [java.lang.UnsatisfiedLinkError]=/tmp/libgpiod17541369362071990391.so: /lib/arm-linux-gnueabihf/libc.so.6: version `GLIBC_2.33' not found (required by /tmp/libgpiod17541369362071990391.so)


Thanks so much for all you help!

**Comment by taartspi** [comment]
1. The first error,  did you leave the slf4j-simple.jar in the POM file ?  I think that jar should be present to prevent this warning.   This needs to be fixed before the next items.  
2. I copied that i2c example you referenced, just changed the device address and it runs.  Please add this line to beginning of the main function  
  System.setProperty("org.slf4j.simpleLogger.defaultLogLevel", "TRACE");     
3. edit the pom and remove the dependency pi4j-plugin-gpiod     This error link ,  not sure why you see it.   Reload the project so that jar is removed and rebuild.   
4.  Run the program and copy the first 50 or so lines of output into this post.]
5. import com.pi4j.io.spi.    sorry, I meant to hover after the spi. to see if the SpiChipSelect is listed.     

I bet the import problem and gpiod link error are somehow related, but so far I sure don't know how.

**Comment by HeroDesign** [comment]
Great thanks so very much for the fantastic support
OK I think I've gotten all the I2C code to run without issue.
Still stuck on the SPI and GPIO code
 1. Yes org.slf4j:slf4j-simple:2.0.16 is in the POM and is loaded 
 2. Added 
 3. I need pi4j-plugin-pigpio for my gpio code, which is also having problems
 5. No SpiChipSelect is not listed. just Spi, SpiConfig, SpiProvider,impl, SpiBase, SpiConfigBuilder, SpiProviderBas
 
4:
SLF4J(W): No SLF4J providers were found.
SLF4J(W): Defaulting to no-operation (NOP) logger implementation
SLF4J(W): See https://www.slf4j.org/codes.html#noProviders for further details.
Exception in thread "main" com.pi4j.library.pigpio.PiGpioException: PIGPIO ERROR: PI_INIT_FAILED; pigpio initialisation failed
	at com.pi4j.library.pigpio.impl.PiGpioBase.validateResult(PiGpioBase.java:265)
	at com.pi4j.library.pigpio.impl.PiGpioBase.validateResult(PiGpioBase.java:251)
	at com.pi4j.library.pigpio.impl.PiGpioNativeImpl.gpioInitialise(PiGpioNativeImpl.java:106)
	at com.pi4j.library.pigpio.PiGpio.initialize(PiGpio.java:159)
	at com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalInputProviderImpl.create(PiGpioDigitalInputProviderImpl.java:72)
	at com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalInputProviderImpl.create(PiGpioDigitalInputProviderImpl.java:43)
	at hero.everbright.hat.IO.setupButtonsAndRelays(IO.java:73)
	at hero.everbright.hat.IO.<init>(IO.java:36)
	at hero.everbright.hat.IO.getInstance(IO.java:44)
	at hero.everbright.brain.EverbrightLive.main(EverbrightLive.java:142)

**Comment by taartspi** [comment]
Did u run the pgm from intellij ?    If yes.  Copy the very long command from the first line in the intellij console. Paste it into a terminal and put sudo at the very start.   Pigpio must run with sudo.   If u use the linuxfs provider the sudo
Requirement is gone.    

https://www.pi4j.com/documentation/providers/

Also. Can u do step 2 and 4.   Assuming step 2 provides the added BoardInfo details
